tap_salesforce/client.py

```diff
@@ -0,0 +1,356 @@
+        """returns the supported table names, as well as the replication_key"""
+        tables = [
+            ("Account", "LastModifiedDate"),
+            ("Contact", "LastModifiedDate"),
+            ("ContactHistory", "CreatedDate"),
+            ("Lead", "LastModifiedDate"),
+            ("Opportunity", "LastModifiedDate"),
+            ("Campaign", "LastModifiedDate"),
+            ("AccountContactRelation", "LastModifiedDate"),
+            ("AccountContactRole", "LastModifiedDate"),
+            ("OpportunityContactRole", "LastModifiedDate"),
+            ("CampaignMember", "LastModifiedDate"),
+            ("OpportunityHistory", "CreatedDate"),
+            ("AccountHistory", "CreatedDate"),
+            ("LeadHistory", "CreatedDate"),
+            ("User", "LastModifiedDate"),
+            ("Invoice__c", "LastModifiedDate"),
+            ("Trial__c", "LastModifiedDate"),
+            ("Task", "LastModifiedDate"),
+        ]
+        table: str
+        replication_key: str
+        for table, replication_key in tables:
+            fields = self.get_fields(table)
+
+            yield (table, fields, replication_key)
+
+    def get_fields(self, table: str) -> Dict[str, Field]:
+        """returns a list of all fields and custom fields of a given table"""
+
+        try:
+            resp = self._make_request(
+                "GET", f"/services/data/{self._API_VERSION}/sobjects/{table}/describe/"
+            )
+
+            sobject = resp.json()
+
+            fields = [
+                Field(name=o["name"], type=o["type"], nullable=o["nillable"])
+                for o in sobject["fields"]
+            ]
+
+            filtered = filter(
+                lambda f: f.type != "json" and f.name not in self._BLACKLISTED_FIELDS,
+                fields,
+            )
+
+            return {f.name: f for f in filtered}
+        except requests.exceptions.HTTPError as err:
+            if err.response is None:
+                raise
+
+            if not err.response.status_code == 404:
+                raise
+
+            return {}
+
+    def get_records(
+        self,
+        table: str,
+        fields: Dict[str, Field],
+        replication_key: Optional[str],
+        start_date: datetime,
+        end_date: Optional[datetime] = None,
+        limit: Optional[int] = None,
+    ) -> Generator[Dict, None, None]:
+        field_names = list(fields.keys())
+
+        select_stm = f"SELECT {','.join(field_names)} "
+        from_stm = f"FROM {table} "
+
+        if replication_key is not None:
+            where_stm = f"WHERE {replication_key} >= {start_date.strftime('%Y-%m-%dT%H:%M:%SZ')} "
+            if end_date:
+                where_stm += f" AND {replication_key} < {end_date.strftime('%Y-%m-%dT%H:%M:%SZ')} "
+
+            order_by_stm = f"ORDER BY {replication_key} ASC "
+        else:
+            where_stm = ""
+            order_by_stm = ""
+
+        if limit:
```

Comment:
if we limit 200 in the query, I think that means even if there are 2000 records, we will only get 200.

Reply:
Ah you probably did it on purpose for testing, so that you don't need to wait for too long.

Reply:
yes, thank you. Should have taken it out for the review.

Reply:
done 👍